browserdebuggertools/wssessionmanager.py

```diff
@@ -15,23 +17,116 @@
 from browserdebuggertools.exceptions import (
     DevToolsException, ResultNotFoundError, TabNotFoundError, MaxRetriesException,
     DevToolsTimeoutException, DomainNotEnabledError,
-    MethodNotFoundError, UnknownError, ResourceNotFoundError, TimerException
+    MethodNotFoundError, UnknownError, ResourceNotFoundError, DeadMessagingThread
 )
 
 
-def open_connection_if_closed(socket_handler_method):
+class _WSMessagingThread(Thread):
+
+    _CONN_TIMEOUT = 15
+    _BLOCKED_TIMEOUT = 5
+    _MAX_QUEUE_BUFFER = 1000
+    _POLL_INTERVAL = 1
+
+    def __init__(self, port, send_queue, recv_queue, poll_signal):
+        super(_WSMessagingThread, self).__init__()
+        self._port = port
+        self._send_queue = send_queue
+        self._recv_queue = recv_queue
+        self._last_poll = None
+        self._continue = True
+        self._poll_signal = poll_signal
+
+        self.exception = None
+        self.ws = self._get_websocket()
+        self.daemon = True
+
+    def __del__(self):
+        self.close()
+
+    def _get_websocket_url(self, port):
+        response = requests.get(
+            "http://localhost:{}/json".format(port), timeout=self._CONN_TIMEOUT
+        )
+        if not response.ok:
+            raise DevToolsException("{} {} for url: {}".format(
+                response.status_code, response.reason, response.url)
+            )
+
+        tabs = [target for target in response.json() if target["type"] == "page"]
+        if not tabs:
+            raise TabNotFoundError("There is no tab to connect to.")
+        return tabs[0]["webSocketDebuggerUrl"]
+
+    def _get_websocket(self):
+        websocket_url = self._get_websocket_url(self._port)
+        logging.info("Connecting to websocket %s" % websocket_url)
+        ws = websocket.create_connection(
+            websocket_url, timeout=self._CONN_TIMEOUT
+        )
+        ws.settimeout(0)  # Don"t wait for new messages
+        return ws
 
-    def retry_if_exception(socket_handler_instance, *args, **kwargs):
+    @contextlib.contextmanager
+    def _ws_io(self):
 
+        # noinspection PyBroadException
         try:
-            return socket_handler_method(socket_handler_instance, *args, **kwargs)
+            yield
+        except Exception as e:
+            self.exception = e
+            logging.exception("WS messaging thread terminated with exception", exc_info=True)
```

Comment:
I will change it to warning level

`logging.warning("WS messaging thread terminated with exception", exc_info=True)`

because we know threads will occasionally die